docs/source/cpp/dataset.rst

```diff
@@ -0,0 +1,381 @@
+
+.. _cpp-dataset-filtering-data:
+
+Filtering data
+--------------
+
+So far, we've read the entire dataset, but this can be inefficient or use too
+much memory. The :class:`arrow::dataset::Scanner` offers control over what data
+to read.
+
+:func:`arrow::dataset::ScannerBuilder::Project` lets us read only specified
+columns:
+
+.. literalinclude:: ../../../cpp/examples/arrow/dataset_documentation_example.cc
+   :language: cpp
+   :lines: 172-191
+   :emphasize-lines: 16
+   :linenos:
+   :lineno-match:
+
+For formats which support pushdown (such as Parquet), this means that only the
+specified columns will be read from the filesystem, saving I/O costs.
+
+A filter can be provided with :func:`arrow::dataset::ScannerBuilder::Filter`,
+so that rows which do not match the filter predicate will not be included in
+the returned table.
+
+.. literalinclude:: ../../../cpp/examples/arrow/dataset_documentation_example.cc
+   :language: cpp
+   :lines: 172-191
+   :emphasize-lines: 17
+   :linenos:
+   :lineno-match:
+
+.. TODO Expressions not documented pending renamespacing
+
+Projecting columns
+------------------
+
+In addition to selecting columns,
+:func:`arrow::dataset::ScannerBuilder::Project` can also be used for more
+complex projections in combination with expressions.
+
+In this case, we pass a vector of expressions used to construct column values,
+and a vector of names for the columns:
+
+.. literalinclude:: ../../../cpp/examples/arrow/dataset_documentation_example.cc
+   :language: cpp
+   :lines: 193-223
+   :emphasize-lines: 18-28
+   :linenos:
+   :lineno-match:
+
+This also determines the column selection; only the given columns will be
+present in the resulting table. If you want to include a derived column in
+*addition* to the existing columns, you can build up the expressions from the
+dataset schema:
+
+.. literalinclude:: ../../../cpp/examples/arrow/dataset_documentation_example.cc
+   :language: cpp
+   :lines: 225-254
+   :emphasize-lines: 17-27
+   :linenos:
+   :lineno-match:
+
+Reading and writing partitioned data
+------------------------------------
+
+So far, we've been working with datasets consisting of flat directories with
+files. However, a dataset can exploit nested directory structures defining a
+partitioned dataset, where sub-directory names hold information about which
+subset of the data is stored in that directory.
+
+For example, a dataset partitioned by year and month may look like on disk:
+
+.. code-block:: text
+
+   dataset_name/
+     year=2007/
+       month=01/
+          data0.parquet
+          data1.parquet
+          ...
+       month=02/
+          data0.parquet
+          data1.parquet
+          ...
+       month=03/
+       ...
+     year=2008/
+       month=01/
+       ...
+     ...
+
+The above partitioning scheme is using "/key=value/" directory names, as found
+in Apache Hive.
+
+Let's create a small partitioned dataset. For this, we'll use Dataset's writing
```

Comment:
It feels a bit weird to have "Dataset" with a capital D and no backticks or link.  Could maybe phrase it as "Arrow's dataset writing functionality."